src/v/container/include/container/contiguous_range_map.h

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "container/fragmented_vector.h"
+
+#include <seastar/util/optimized_optional.hh>
+
+#include <concepts>
+#include <iterator>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+/**
+ * Contiguous range map is an associative sorted container backed by
+ * chunked_vector designed to efficiently store objects indexed with contiguous
+ * and limited range of integers. The container provides a wrapper around basic
+ * chunked vector that reassembles API of a standard map.
+ *
+ * The wrapper allows random inserts of elements to the map event tho this is
+ * not supported by standard vector.
+ *
+ * Underlying container is resized every time its size is not sufficient to
+ * account for emplaced key.
+ *
+ * The contiguous_range_map tolerates gap in the range of keys however number
+ * and size of gaps is proportional to performance penalty hit when incrementing
+ * or decrementing iterators.
+ */
+template<std::integral KeyT, typename ValueT>
+class contiguous_range_map {
+public:
+    using value_type = std::pair<const KeyT, ValueT>;
+    using key_type = KeyT;
+    using mapped_type = ValueT;
+
+private:
+    using underlying_t = chunked_vector<std::optional<value_type>>;
```

Comment:
what is the use of making it chunked_vector<std::optional<vaue_type>> instead of chunked_vector<std::optional\<ValueT\>\>, for iterator we already have access to the index so we can construct the pair on the fly?

Reply:
we would always need a copy then. I was considering that but the API wasn't ergonomic and it was differnt than for the standard map